### PR TITLE
[EMB-341] Personal access tokens page

### PR DESCRIPTION
**Pull request by aaxelb**
## Purpose
Add the personal access tokens tab of the settings page.

## Summary of Changes
- Add `token`, `scope` model, adapter, serializer
  - Add `osf-model` blueprint to make that easier in the future
  - Update `osf-serializer` to always include attributes with `serialize: true` in the serializer's `attrs`
- New routes:
  - `settings`
  - `settings.tokens`
  - `settings.tokens.edit`
  - `settings.tokens.create`
- `paginated-list/*` - family of components for all sorts of paginated lists
  - rename `paginated-relation` to `paginated-list/has-many`
  - add `paginated-list/all` for lists of all of a type of model
- `validated-input/*` - split `validated-input` into a family of components, since each type has a different interface and behavior
  - `validated-input/checkbox`
  - `validated-input/checkboxes` (new!)
  - `validated-input/date`
  - `validated-input/recaptcha`
  - `validated-input/text`
  - `validated-input/textarea`
  

## Side Effects
Things I touched while deduplicating logic (should all work the same):
- Sign-up form on the logged-out landing page
- Quick files "share" buttons (on file browser and file detail page)
- Paginated lists:
  - Node/registration forks page
  - Node/registration analytics page "linked nodes" list

## Feature Flags

- `ember_user_settings_tokens_page`: enable the embosf'd tokens page
## QA Notes
The tokens page should stay waffled off for now, there are a few cosmetic changes incoming in another PR.

## Ticket

https://openscience.atlassian.net/browse/EMB-341

# Reviewer Checklist

- [ ] meets requirements
- [ ] easy to understand
- [ ] DRY
- [ ] testable and includes test(s)
- [ ] changes described in `CHANGELOG.md`

### CODE REVIEW

File: app/authenticators/osf-cookie.ts [hunk 1]
```diff
@@ -41,7 +42,7 @@ export default class OsfCookie extends Base {
             url: `${apiUrl}/${apiNamespace}/`,
         });
 
-        if (devMode) {
+        if (devMode && !Ember.testing) {
```
Comment (jamescdavis) [inline app/authenticators/osf-cookie.ts]: Can probably remove.

File: app/serializers/token.ts [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+            newHash = {
+                ...hash,
+                relationships: {
+                    ...hash.relationships,
+                    scopes: {
+                        data: scopes.split(' ').map((s: string) => ({ id: s, type: 'scopes' })),
+                    },
+                },
+            };
+        }
```
Comment (jamescdavis) [inline app/serializers/token.ts]: Fix the back end so we don't have to do this.

File: app/settings/tokens/index/route.ts [hunk 3]
```diff
@@ -0,0 +1,4 @@
+import Route from '@ember/routing/route';
+
+export default class SettingsTokensIndexRoute extends Route {
+}
```
Comment (jamescdavis) [inline app/settings/tokens/index/route.ts]: Don't need this file.

File: lib/osf-components/addon/components/sign-up-form/template.hbs [hunk 4]
```diff
@@ -1,37 +1,36 @@
...
+    {{validated-input/text password=true
         model=model
-        didValidate=didValidate
         valuePath='password'
-        placeholder=(t 'sign_up_form.password')
-        key-up=(perform strength value=model.password)
+        messagesShown=didValidate
         disabled=hasSubmitted
-        aria-label=(t 'sign_up_form.password')
+        onKeyUp=(perform strength model.password)
```
Comment (jamescdavis) [inline lib/osf-components/addon/components/sign-up-form/template.hbs]: make sure this shouldn't be `keyUp`

File: lib/osf-components/addon/components/tos-consent-banner/template.hbs [hunk 5]
```diff
@@ -15,14 +15,13 @@
                     }}
                 </h4>
                 <h5>
-                    {{#validated-input
-                        type='checkbox'
+                    {{#validated-input/checkbox
                         model=currentUser.user
-                        didValidate=didValidate
                         valuePath='acceptedTermsOfService'
+                        didValidate=didValidate
```
Comment (jamescdavis) [inline lib/osf-components/addon/components/tos-consent-banner/template.hbs]: should be `messagesShown=didValidate`

File: lib/osf-components/addon/components/project-selector/template.hbs [hunk 6]
```diff
@@ -15,13 +15,13 @@
...
+            {{validated-input/text
                 model=newProject
-                didValidate=didValidate
                 valuePath='title'
+                didValidate=didValidate
                 placeholder=(t 'move_to_project.enter_project_title')
-                aria-label=(t 'move_to_project.enter_project_title')
-                key-up=(action 'valueChanged')
+                ariaLabel=(t 'move_to_project.enter_project_title')
+                keyUp=(action 'valueChanged')
```
Comment (jamescdavis) [inline lib/osf-components/addon/components/project-selector/template.hbs]: is this right?

File: lib/osf-components/addon/components/project-selector/template.hbs [hunk 7]
```diff
@@ -15,13 +15,13 @@
 {{#if (eq projectSelectState 'newProject')}}
     <form onsubmit={{action 'prevent'}}>
         <div class="form-group">
-            {{validated-input
+            {{validated-input/text
                 model=newProject
-                didValidate=didValidate
                 valuePath='title'
+                didValidate=didValidate
```
Comment (jamescdavis) [inline lib/osf-components/addon/components/project-selector/template.hbs]: should be messagesShown

File: lib/osf-components/addon/components/validated-input/base-component.ts [hunk 8]
```diff
@@ -0,0 +1,70 @@
...
+
+    // Private properties
+    validation?: any; // defined in constructor
+    value?: any; // defined in constructor
+
+    @className('has-success')
+    @equal('validationStatus', ValidationStatus.Success)
+    hasSuccess!: boolean;
+
+    // TODO delete?
```
Comment (jamescdavis) [inline lib/osf-components/addon/components/validated-input/base-component.ts]: look into this

File: lib/osf-components/addon/components/validated-input/checkboxes/x-checkbox/component.ts [hunk 9]
```diff
@@ -0,0 +1,35 @@
+import { tagName } from '@ember-decorators/component';
+import { computed } from '@ember-decorators/object';
+import Component from '@ember/component';
+
+import { localClassNames } from 'ember-osf-web/decorators/css-modules';
+import styles from './styles';
+import layout from './template';
+
+@tagName('')
+@localClassNames('ValidatedInputCheckboxesXCheckbox')
```
Comment (jamescdavis) [inline lib/osf-components/addon/components/validated-input/checkboxes/x-checkbox/component.ts]: remove this and associated stuff

File: lib/osf-components/addon/components/validated-input/checkboxes/x-checkbox/component.ts [hunk 10]
```diff
@@ -0,0 +1,35 @@
...
+export default class ValidatedInputCheckboxesXCheckbox extends Component {
+    layout = layout;
+    styles = styles;
+
+    // Required arguments
+    relationArray!: any[];
```
Comment (jamescdavis) [inline lib/osf-components/addon/components/validated-input/checkboxes/x-checkbox/component.ts]: make generic

File: lib/osf-components/addon/components/validated-input/checkboxes/x-checkbox/template.hbs [hunk 11]
```diff
@@ -0,0 +1,10 @@
+{{input
+    (html-attributes aria-label=this.ariaLabel)
+    type='checkbox'
+    checked=this.checked
+    id=this.checkboxId
+    name=this.checkboxName
+    disabled=this.disabled
+}}
+
+{{yield this.checkboxId}}
```
Comment (jamescdavis) [inline lib/osf-components/addon/components/validated-input/checkboxes/x-checkbox/template.hbs]: probably not necessary

File: lib/osf-components/app/components/paginated-list/all/component.ts [hunk 12]
```diff
@@ -0,0 +1 @@
+export { default } from 'osf-components/components/paginated-list/all/component';
```
Comment (jamescdavis) [inline lib/osf-components/app/components/paginated-list/all/component.ts]: js

File: mirage/models/root.js [hunk 13]
```diff
@@ -1,5 +1,15 @@
...
+    featureFlagNames: {
+        routes,
+    },
+} = config;
 
 export default Model.extend({
     currentUser: belongsTo('user'),
+
+    // Pretend all routes are flagged on
+    activeFlags: Object.values(routes),
```
Comment (jamescdavis) [inline mirage/models/root.js]: rm (in factory)

File: app/serializers/token.ts (hunk 2 above)
Comment (aaxelb) [reply app/serializers/token.ts]: Made [a ticket](https://openscience.atlassian.net/browse/PLAT-1062)

File: app/locales/en/translations.ts [hunk 14]
```diff
@@ -38,6 +38,7 @@ export default {
         ellipsis: '\u2026',
         warning: 'Warning',
         clipboard_copy: 'Copy to clipboard',
+        clipboard_copied: 'Copied!',
```
Comment (jamescdavis) [inline app/locales/en/translations.ts]: consider moving to `osf-components` section

File: lib/osf-components/addon/components/file-browser/template.hbs [hunk 15]
```diff
@@ -5,7 +5,7 @@
                 (html-attributes aria-label=(t 'file_browser.filtering'))
                 placeholder=(t 'file_browser.filtering')
                 class="form-control"
-                keyUp=(action updateFilter value='target.value')
+                key-up=(action updateFilter value='target.value')
```
Comment (jamescdavis) [inline lib/osf-components/addon/components/file-browser/template.hbs]: double check both work (but should go with what the docs say regardless)

File: lib/osf-components/addon/components/paginated-list/base-data-component.ts [hunk 16]
```diff
@@ -10,8 +10,12 @@ import defaultTo from 'ember-osf-web/utils/default-to';
 export default abstract class BaseDataComponent extends Component.extend({
     loadItemsWrapperTask: task(function *(this: BaseDataComponent) {
         const blocker = this.ready.getBlocker();
+
+        // Resolve race condition on init: Let component finish initializing before continuing
+        yield;
```
Comment (jamescdavis) [inline lib/osf-components/addon/components/paginated-list/base-data-component.ts]: add TODO for removal once task decorators

File: lib/osf-components/addon/components/paginated-list/has-many/component.ts [hunk 17]
```diff
@@ -0,0 +1,83 @@
...
+    }).restartable();
+
+    @or('model', 'modelTaskInstance.value')
+    modelInstance?: OsfModel;
+
+    constructor(...args: any[]) {
+        super(...args);
+
+        assert(
+            'Must provide either `model` xor `modelTaskInstance` to {{paginated-relation}}',
```
Comment (jamescdavis) [inline lib/osf-components/addon/components/paginated-list/has-many/component.ts]: `s/paginated-relation/paginated-list\/has-many/`

File: lib/osf-components/addon/components/paginated-list/layout/template.hbs [hunk 18]
```diff
@@ -0,0 +1,37 @@
+{{#if this.loading}}
+    {{#if this.placeholderCount}}
+        <ul class="list-group m-md">
+            {{#each (range 1 @placeholderCount)}}
```
Comment (jamescdavis) [inline lib/osf-components/addon/components/paginated-list/layout/template.hbs]: Should be `this.placeholderCount`

Comment (aaxelb) [reply lib/osf-components/addon/components/paginated-list/layout/template.hbs]: I knew I never should have trusted the `@`...

**Pull request merged**